File: scalaGemMigrationExamples/gems/src/main/scala/rohitjainprophecyioteam/scalagemsmigrationexamples/gems/ReformatCustom.scala
package rohitjainprophecyioteam.scalagemsmigrationexamples.gems

import io.prophecy.gems._
import io.prophecy.gems.componentSpec.{ColumnsUsage, ComponentSpec}
import io.prophecy.gems.copilot.{ProjectionExpression, SelectStmt}
import io.prophecy.gems.dataTypes._
import io.prophecy.gems.diagnostics._
import io.prophecy.gems.uiSpec._
import org.apache.spark.sql.{DataFrame, SparkSession}
import play.api.libs.json.{Json, OFormat}

object Reformat extends ComponentSpec {

  val name: String = "ReformatCustom"
  val category: String = "Transform"
  val gemDescription: String = "Edits column names or values using expressions."
  val docUrl: String = "https://docs.prophecy.io/low-code-spark/gems/transform/reformat/"

  type PropertiesType = ReformatProperties

  override def optimizeCode: Boolean = true

  // todo: columnSelector property can be hidden from user
  // and move column select to the ports
  case class ReformatProperties(
    @Property("Columns selector")
    columnsSelector: List[String] = Nil,
    @Property("Column expressions", "List of all the column expressions")
    expressions: List[SColumnExpression] = Nil,
    @Property("new property")
    newProperty: Option[String] = None,
  ) extends ComponentProperties

  implicit val reformatPropertiesFormat: OFormat[ReformatProperties] = Json.format[ReformatProperties]

  def dialog: Dialog = Dialog("Reformat")
    .addElement(
      ColumnsLayout(gap = Some("1rem"), height = Some("100%"))
        .addColumn(
          PortSchemaTabs(
            allowInportRename = true,
            selectedFieldsProperty = Some("columnsSelector"),
            singleColumnClickCallback = Some { (portId: String, column: String, state: Any) ⇒
              val st = state.asInstanceOf[Component]
              val existingTargetNames = st.properties.expressions.map(_.target)
              val targetCol =
                getTargetTokens(column, existingTargetNames.map(_.split('.').toList), shortest = true).mkString(".")
              st.copy(properties =
                st.properties.copy(
                  expressions = st.properties.expressions ::: List(
                    SColumnExpression(targetCol, s"""col("${sanitizedColumn(column)}")""")
                  )
                )
              )
            },
            allColumnsSelectionCallback = Some { (portId: String, state: Any) ⇒
              val st = state.asInstanceOf[Component]
              val columnsInSchema = getColumnsInSchema(portId, st, SchemaFields.TopLevel)
              val updatedExpressions = st.properties.expressions ::: columnsInSchema.map { x ⇒
                SColumnExpression(x, s"""col("${sanitizedColumn(x)}")""")
              }
              st.copy(properties = st.properties.copy(expressions = updatedExpressions))
            }
          ).importSchema(),
          "2fr"
        )
        .addColumn(
          ExpTable("Reformat Expression")
            .withRowId()
            .bindProperty("expressions")
            .withCopilotEnabledExpressions(
              CopilotSpec(
                method = "copilot/getExpression",
                methodType = Some("CopilotProjectionExpressionRequest"),
                copilotProps = CopilotPromptTypeProps(buttonLabel = "Ask AI")
              )
            ),
          "5fr"
        )
    )

  def validate(component: Component)(implicit context: WorkflowContext): Diagnostics = {
    val diagnostics = validateExpTable(
      component.properties.expressions,
      "expressions",
      component,
      testColumnPresence = Some(ColumnsUsage.WithoutInputAlias)
    )
    diagnostics.toList
  }

  def onChange(oldState: Component, newState: Component)(implicit
    context: WorkflowContext
  ): Component = {
    val newProps = newState.properties

    val usedColExps = getColumnsToHighlight(newProps.expressions, newState)

    // todo: column selector internal format must be hidden from user.
    // move following logic in ComponentSpec trait
    newState.copy(properties =
      newProps.copy(
        columnsSelector = usedColExps,
        expressions = newProps.expressions.map(_.withRowId)
      )
    )
  }

  override def getCPStmt(component: Component): Option[SelectStmt] = {
    val props = component.properties
    Some(
      SelectStmt(
        component.component,
        props.expressions.map(x => ProjectionExpression(Some(x.target), x.expression.expression)),
        Nil,
        Nil,
        Nil
      )
    )
  }

  override def userReadme: String =
    """
      |## Reformat
      |---
      |
      |Reformat is often use to edit one or more columns' values, often by simple expressions and functions. Often this involves expressions such as `substring` to extract parts of incoming string, `case when` statement to add columns summarizing conditions of incoming data. This is also used to rename columns and select a few columns.
      |
      |> **Note:** Any columns not selected explicitly are removed, use `Select All` in the left bar to add all columns by default.
      |
      |When there are no columns selected, all columns are passed through to the output. However, once you start selecting some columns, only the selected columns are output.
      |
      |> **Note:** _Implementation_: Reformat converts to a SQL `Select` or in relational terms into a `projection`. However, when changing (add, edit, delete) one or two columns, one can use [SchemaTransformer](https://docs.prophecy.io/spark/standard_components/transforms/reformat/) that uses underlying `withColumn` construct. However, Reformat is best for performance. For more details refer to [Spark APIs](https://spark.apache.org/docs/latest/api/scala/org/apache/spark/sql/Dataset.html)
      |
      |
      |
      |## Example
      |---
      |
      |Reformat gives you a spreadsheet like interface where you can write expressions based on incoming columns to produce new columns.
      |
      |___Expression Builder___
      |
      |There is an `expression builder` that enables one to quickly write expressions. Expression builder will suggest
      |
      |- Incoming columns
      |- Inbuilt operators
      |- In-built functions
      |- User defined functions
      |
      |
      |## Code
      |---
      |
      |This component is a simple select statement, however the expressions can become very large and we see tables of 1000+ columns frequently with this component being hundreds of lines long.
      |
      |
      |```scala
      |object PrepareComponent {
      |
      |    def apply(spark: SparkSession, in: DataFrame): Reformat = {
      |        import spark.implicits._
      |
      |        val out = in.select(
      |          datediff(current_date(), col("account_open_date")).as("account_length_days"),
      |          col("order_id"),
      |          col("customer_id"),
      |          col("amount"),
      |          col("first_name"),
      |          col("last_name"),
      |          col("phone"),
      |          col("email"),
      |          col("country_code"),
      |          col("account_flags")
      |        )
      |
      |        out
      |    }
      |}
      |```
      |""".stripMargin

  class ReformatCode(props: PropertiesType)(implicit context: WorkflowContext) extends ComponentCode {

    def apply(spark: SparkSession, in: DataFrame): DataFrame = {
      val out = props.expressions match {
        case Nil ⇒ in
        case _ ⇒ in.select(props.expressions.columns: _*)
      }
      out
    }

  }

  override def deserializeProperty(props: String): ReformatProperties =
    Json.parse(props).as[ReformatProperties]

  override def serializeProperty(props: ReformatProperties): String =
    Json.stringify(Json.toJson(props))

  registerPropertyEvolution(AddNewProperty)
}

File: PythonGemsMigrationExamples/gems/rohitjainprophecyioteam_pythongemsmigrationexamples/gems/ReformatCustom.py
from prophecy.cb.server.base.ComponentBuilderBase import *
from pyspark.sql import *
from pyspark.sql.functions import *
import re

from prophecy.cb.ui.UISpecUtil import getColumnsToHighlight, computeTargetName, SchemaFields, getColumnsInSchema, \
    validateExpTable, ColumnsUsage, getTargetTokens, sanitizedColumn
from prophecy.cb.util.CSVUtils import parse_escaped_csv, unparse_escaped_csv, CSVParseException
from prophecy.cb.ui.uispec import *
from prophecy.cb.util.StringUtils import isBlank
from prophecy.cb.server.base import WorkflowContext

class ReformatCustom(ComponentSpec):
    name: str = "ReformatCustom"
    category: str = "Transform"
    gemDescription: str = "Edits column names or values using expressions."
    docUrl: str = "https://docs.prophecy.io/low-code-spark/gems/transform/reformat/"



    def optimizeCode(self) -> bool:
        return True

    @dataclass(frozen=True)
    class ReformatProperties(ComponentProperties):
        columnsSelector: List[str] = field(default_factory=list)
        expressions: List[SColumnExpression] = field(default_factory=list)
        activeTab: str = "expressions"
        importLanguage: str = "${$.workflow.metainfo.frontEndLanguage}"
        importString: str = ""

    def onClickFunc(self, portId: str, column: str, state: Component[ReformatProperties]):
        existingTargetNames = list(map(lambda exp: exp.target, state.properties.expressions))
        targetTokens = getTargetTokens(column, [exp.split('.') for exp in existingTargetNames], True)
        targetCol = '.'.join(targetTokens)
        expressions = state.properties.expressions
        expressions.append(SColumnExpression(targetCol, SColumn.getSColumn(sanitizedColumn(column)), ""))
        return state.bindProperties(replace(state.properties, expressions=expressions))

    def allColumnsSelectionFunc(self, portId: str, state: Component[ReformatProperties]):
        columnsInSchema = getColumnsInSchema(portId, state, SchemaFields.TopLevel)
        expressions = list(
            map(lambda column: SColumnExpression.getSColumnExpression(sanitizedColumn(column)), columnsInSchema))
        state.properties.expressions.extend(expressions)
        return state.bindProperties(replace(state.properties, expressions=state.properties.expressions))

    def expressions_to_csv(self, state: Component[ReformatProperties]) -> Component[ReformatProperties]:
        exprCSV = []
        for e in state.properties.expressions:
            linepart = [e.target, e.expression.rawExpression]
            if len(e.description):
                linepart.append(e.description)
            exprCSV.append(linepart)
        csv_string = unparse_escaped_csv(exprCSV)
        return state.bindProperties(replace(state.properties, importString=csv_string))

    def csv_to_expressions(self, state: Component[ReformatProperties]) -> Component[ReformatProperties]:
        exprs = []
        # for (target, exp, desc) in parse_escaped_csv(state.properties.importString):
        for line in parse_escaped_csv(state.properties.importString, field_min=2, field_max=2):
            (target, exp) = line
            exprs.append(SColumnExpression(
                target.strip(),
                SColumn(
                    exp,
                    state.properties.importLanguage,
                    col(exp),
                    [exp]
                ),
                ""
            ))
        return state.bindProperties(replace(state.properties, expressions=exprs))

    def dialog(self) -> Dialog:
        expTable = ExpTable("Reformat Expression") \
            .enableVirtualization() \
            .bindProperty("expressions") \
            .withCopilotEnabledExpressions(
                CopilotSpec(
                    method="copilot/getExpression",
                    methodType="CopilotProjectionExpressionRequest",
                    copilotProps=CopilotPromptTypeProps(
                        buttonLabel="Ask AI",
                    )
                )
            )
        bulkEdit = StackLayout(height="100%") \
            .addElement(
            NativeText("Edit the Reformat expressions in the field below. Use the format of \"name,expr\".")
        ).addElement(
            NativeText("Use ``...`` to wrap multi-line expressions.")
        ).addElement(
            Editor(height="100%", language="${component.properties.importLanguage}") \
                .bindProperty("importString")
        )
        refTabs = Tabs() \
            .bindProperty("activeTab") \
            .addTabPane(
            TabPane("Expressions", "expressions").addElement(expTable)
        ).addTabPane(
            TabPane("Advanced", "advanced").addElement(bulkEdit)
        )
        return Dialog("Reformat").addElement(
            ColumnsLayout(gap="1rem", height="100%")
                .addColumn(
                PortSchemaTabs(
                    allowInportRename=True,
                    selectedFieldsProperty="columnsSelector",
                    singleColumnClickCallback=self.onClickFunc,
                    allColumnsSelectionCallback=self.allColumnsSelectionFunc
                ).importSchema(),
                "2fr"
            )
                .addColumn(refTabs, "5fr")
        )

    def validate(self, context: WorkflowContext, component: Component[ReformatProperties]) -> List[Diagnostic]:
        diagnostics = []
        expTableDiags = validateExpTable(component.properties.expressions, "expressions", component,
                                         ColumnsUsage.WithoutInputAlias)
        diagnostics.extend(expTableDiags)

        if component.properties.activeTab == "advanced":
            try:
                parse_escaped_csv(component.properties.importString, field_min=2, field_max=3)
            except CSVParseException as e:
                diagnostics.append(Diagnostic(f"properties.importString", str(e), SeverityLevelEnum.Error))

        return diagnostics

    def onChange(self, context: WorkflowContext, oldState: Component[ReformatProperties], newState: Component[ReformatProperties]) -> Component[
        ReformatProperties]:

        oldProps = oldState.properties
        newProps = newState.properties

        if oldProps.activeTab == "advanced" and newProps.activeTab == "expressions":
            try:
                newState = self.csv_to_expressions(newState)
                newProps = newState.properties

            except CSVParseException:
                pass
        elif oldProps.activeTab == "expressions" and newProps.activeTab == "advanced":
            newState = self.expressions_to_csv(newState)
            newProps = newState.properties

        expressions = newProps.expressions

        usedColExps = getColumnsToHighlight(expressions, newState)

        return newState.bindProperties(replace(newProps,
                                               columnsSelector=usedColExps,
                                               expressions=list(
                                                   map(lambda exp: exp.withRowId(), expressions))))

    def getCPStmt(self, component: Component[ReformatProperties]) -> Optional[SelectStmt]:
        return SelectStmt(
            component.component,
            [ProjectionExpression(expression.target, expression.expression.rawExpression) for expression in
             component.properties.expressions]
        )

    class ReformatCode(ComponentCode):
        def __init__(self, newProps):
            self.props: Reformat.ReformatProperties = newProps

        def apply(self, spark: SparkSession, in0: DataFrame) -> DataFrame:
            if len(self.props.expressions) > 0:
                selectColumns = map(lambda x: x.column(), self.props.expressions)
                return in0.select(*selectColumns)
            else:
                return in0